### ISSUE TITLE
Nifty ListBox: Custom Item and Selection

### ISSUE THREAD

**Issue by Sekagra** [nifty-gui/nifty-gui#132]
I created a Nifty ListBox with a custom ListItem control, containing a few text elements as well as an image.
 With the help of this Tutorial [here](http://sourceforge.net/apps/mediawiki/nifty-gui/index.php?title=Standard_Controls_ListBox) this was easy doable, so thanks for that.

``` xml
 <control id="shopBuyList" name="listBox" vertical="off" horizontal="off" selection="Single" viewConverterClass="de.ethasia.terratown.gui.ShopItemConverter">
 <control name="shopItem" />
 </control>
```

However I encountered a few issues when it comes to selecting items in my new ListBox. As the Tutorial notes in the last part, I have to create a style, or at least assign an existing one. I just assumed that the style has to be set on the ListItem control I created, as the Tutorial doesn’t mention anything specifically.

With the standard 'nifty-listbox-item'-Style, the color changes just fine when hovering above an item, **but I can't select an item by clicking on it**. (The defined 'Select' color is not applied and onListBoxSelectionChanged doesn't fire).
The only thing happening after clicking a random item in my list is, that the hover-color is applied on the **first item**. However I can use 'Enter' to get at least the **first item** selected.

I’m certain that this has only to do with my custom Item and/or the style, because if I remove it from the ListBox and just add strings, everything behaves as expected.

I also created a custom style, changing the default one Nifty uses in a way that is at least logical to me, by setting the onClick-Effect to 'select':

``` xml
<style id="shop-listbox-item" base="nifty-label">
  <attributes color="#000f" width="100%" align="left" textVAlign="center" textHAlign="left" />
  <interact onClick="listBoxItemClicked()" />
  <effect>
    <onCustom customKey="focus" name="colorBar" post="false" color="#361db6" neverStopRendering="true" timeType="infinite" />
    <onCustom customKey="select" name="colorBar" post="false" color="#c93939" neverStopRendering="true" timeType="infinite" />
    <onHover name="focus" targetElement="#parent#parent" />
    <onClick name="select" targetElement="#parent#parent" />
  </effect>
</style>
```

(I also tried changing the targetElement paths.)

**Comment by void256** [comment]
I think the problem is that the control you set as the listbox item is missing an actual controller. That way the `interact onClick="listBoxItemClicked()"` used in the style does not find a method to call. That method does the actual triggering of the "select" effect and so on.

I think this can be fixed with something like:

``` java
...
<control name="shopItem" controller="de.lessvoid.nifty.controls.listbox.ListBoxItemController" />
...
```

Or you can set your own controller class but then you have to provide a "listBoxItemClicked()" method similar to the one found in the original class: https://github.com/void256/nifty-gui/blob/1.3/nifty-controls/src/main/java/de/lessvoid/nifty/controls/listbox/ListBoxItemController.java

I've modified the wiki page (which I think you refered to when you said "tutorial") and added the missing controller part: http://sourceforge.net/apps/mediawiki/nifty-gui/index.php?title=Standard_Controls_ListBox

Hope that helps getting it to work...

**Comment by Sekagra** [comment]
Thank you very much for your quick answer. It did indeed help.

In fact I've had already tried using a Controller for the Custom Item and was able to get the _listBoxItemClicked()_-Method to run, but I discarded that solution as it didn't work out with the selection as well. (Which seems logical now that I see the code of the standard implementation of listBoxItemClicked...)

So now there is quite an improvement after I assigned the default _ListBoxItemController_ like you suggested. In combination with the following style I can actually click Items now and get (unfortunately still the first one) selected this way. 

``` xml
<style id="shop-listbox-item-default" base="nifty-label">
  <attributes color="#000f" width="100%" align="left" textVAlign="center" textHAlign="left" />
  <interact onClick="listBoxItemClicked()" />
  <effect>
    <onCustom customKey="focus" name="colorBar" post="false" color="#be3f" neverStopRendering="true" timeType="infinite" />
    <onCustom customKey="select" name="colorBar" post="false" color="#444f" neverStopRendering="true" timeType="infinite" />
    <onCustom customKey="select" name="textColor" post="false" color="#fc0f" neverStopRendering="true" timeType="infinite" />
    <onHover name="colorBar" color="#1e3f" post="false" neverStopRendering="true" timeType="infinite" inset="1px"/>
    <onClick name="focus" targetElement="#parent#parent"/>
  </effect>
</style>
```

So as I mentioned the only problem that remains is **that I still can only get the first item selected.** This however may not be an issue related to the style problem, Because that also was an issue previously, when I used the arrow keys and 'Return' to select something?

(Note: I did meant the Wiki when I referred to the Tutorial, just fixed the link in my op.)

**Comment by void256** [comment]
Could it be that you add the same instance of your own model class several times to the ListBox?

Internally Nifty will use `list.contains(item)` to check things and if your own model class does not correctly implement "equals" Nifty could be confused, I think.

Besides that I can't think of something that would cause that. Maybe you can share your Model class and the control code for "shopItem" too?

**Comment by Sekagra** [comment]
Sure anything that helps.
During my research on the problem I stumbled across an issue reporting problems similar to mine when a single item was added multiple times. However I can definitely rule that out. This happens in my ScreenController:

``` java
Tool stoneHoe = new Tool(...);
Tool copperHoe = new Tool(...); 

this.shopBuyList.addItem(stoneHoe);
this.shopBuyList.addItem(copperHoe);
```

**Tool** is a sub class of **Item** which is what I wrote the ListBoxViewConverter for.
It sets the text of three text elements and one nifty image based on the given Item.

``` java
public void display(Element listBoxItem, Item shopItem) {

    this.shopItemNameText = listBoxItem.findElementByName(ITEMNAME_TEXT_NAME);
    this.shopItemDescriptionText = listBoxItem.findElementByName(DESCRIPTION_TEXT_NAME);      
    this.shopItemPriceText = listBoxItem.findElementByName(PRICE_TEXT_NAME);
    this.shopItemIconImage = listBoxItem.findElementByName(ICON_IMAGE_NAME);

    this.shopItemNameText.getRenderer(TextRenderer.class).setText(shopItem.getName());
    this.shopItemPriceText.getRenderer(TextRenderer.class).setText(String.valueOf(shopItem.getPrice()));

    Nifty nifty = listBoxItem.getNifty();
    NiftyImage iconImage = nifty.getRenderEngine().createImage(shopItem.getIconPath(), false);
    shopItemIconImage.getRenderer(ImageRenderer.class).setImage(iconImage);

    listBoxItem.layoutElements();
}

public int getWidth(Element elmnt, Item t) {
    return 200;
}
```

As mentioned, I applied the default ListBoxItemController to my custom control "shopItem". So there is no further code for that. 
My model class **Item** is nothing special, it doesn't extend anything else and is just returning some string values through getXYZ()-Methods as seen in the Converter.

Besides that, the only "code" involved would be the definition of my shopItem Control:

``` xml
<controlDefinition name="shopItem" style="shop-listbox-item-default">
  <panel childLayout="horizontal" width="100%" align="left" id="shopItemPanel">
    <image id="shopItemIcon" width="50px" height="50px" imageMode="normal" />
    <panel childLayout="vertical" width="300px">
      <text id="shopItemName" style="base-font" color="#00FFFFFF" text="Item Name" textHAlign="left" />
      <text id="shopItemDescription" style="base-font" color="#00FFFFFF" text="Description of this Item." wrap="true" width="210px" textHAlign="left" />
      <text id="shopItemPrice" style="base-font" color="#00FFFFFF" text="Price" textHAlign="left" />
    </panel>
  </panel>
</controlDefinition>
```

I hope that gives any additional information.

**Comment by void256** [comment]
Well, the only thing that looks a bit wrong is the use of ids in your shopItem control. Since ids should be unique in a Nifty screen you'll get conflicts when you add multiple shopItems that share the same ids for their child elements. For instance if you add two shopItems you would end up with two elements on your Screen sharing the same id "shopItemName" for example. I don't think this will cause the selection problem though but it would give you a warning in the log. It would be better to use the "#yourid" syntax to prevent the warning.

Nothing else seems really wrong in the remaining code. Don't know. Maybe send a SSCCE with all classes concerning the problem so that I can reproduce your problem - which in fact I've tried using the code parts you provided with not much luck so far :frowning: Maybe it's the combination of your classes or something unrelated that's not obvious from the code snippets you provided.

PS: I've build you a Listbox example using your code samples. You can download it here: https://dl.dropboxusercontent.com/u/107560313/nifty/issue-132.zip

Unzip it and then build and run it with these two simple commands (Maven 3.x required though. When you are on Mac use a 1.6 JRE. Other platforms should be ok with any JRE):

```
mvn clean package
mvn exec:java -Dexec.mainClass="de.lessvoid.issues.issue132.IssueMain"
```

This works for me and shows a ListBox with three ShopItem entries that I can really select here. I swear! :wink:

**Comment by Sekagra** [comment]
Wow you made quite an effort for this. I really appreciate that.
I'm glad to tell you that I actually solved my problem now and.. that it was caused by an utterly dumb mistake. 
While I was rebuilding my nifty window, testing your code, I realized that I had already created a custom equals()-method in my **Item** class. When I created the test instances (Items) I was under the assumption that the values (Text, Description etc.) would be totally irrelevant and actually gave two elements the same ID (copy&paste). However, equals() was already coded to compare the ID attributes and therefore delivered true in this case.

So everything works fine now. Thanks again.
